### DISCUSSION TITLE
Should `autoquit` be enabled by default?

### DISCUSSION THREAD

**Discussion by joelim-work** [gokcehan/lf#1767] [Polls]
Currently the `autoquit` option (server automatically quits when there are no clients left) is disabled by default.

However it has been suggested to enable it by default, see https://github.com/gokcehan/lf/issues/912#issuecomment-1221303371 and https://github.com/gokcehan/lf/issues/1761#issuecomment-2188046128.

---

Pros:

- New users may be surprised to see a server instance of `lf` being left behind, and enabling `autoquit` by default prevents that from happening.
- This is similar to how the `tmux` server works by default (set using the `exit-empty` option).

Cons:

- This is a breaking change, and may require users to update their configuration.
- Some users may not want to start the server every time `lf` is started, especially if it causes delays.

**Comment by ilyagr** [comment]
Everyone should try inserting `set autoquit` in their config and see if they notice any problems.

-------------

I don't have a clear opinion, but it turns out that setting `autoquit` does affect my config by making a bug I had in it worse. The bug is https://github.com/gokcehan/lf/issues/1293.

I haven't spent the time to read carefully about the workarounds, but I didn't immediately find one. Does anyone have a suggestion for how to avoid this?

## Details

The problematic snippet seems to be:

```
cmd recol ${{
    w=$(tput cols)
    if [ $w -le 30 ]; then
        set nopreview
        lf -remote "send $id set ratios 1"
    elif [ $w -le 50 ]; then
        lf -remote "send $id set ratios 1:1"
    elif [ $w -le 65 ]; then
        lf -remote "send $id set ratios 3:4"
    elif [ $w -le 95 ]; then  # Was 160
        lf -remote "send $id set ratios 2:3:4"
        set preview
    else
        lf -remote "send $id set ratios 2:2:3:4"
        set preview
    fi
}}

recol
```

On the first run of `lf`, the `recol` command results in this message:

```console
$ lf
listen: send: no such client id is connected
```

Without `autoquit`, this only happens once. If I put `set autoquit` in the config, this happens every time `lf` is started (if the server isn't running).

**Reply by DusanLesan** [reply]
> Does anyone have a suggestion for how to avoid this?

@ilyagr Did you try using `on-redraw`? I use it for the same use case without issues

**Reply by ilyagr** [reply]
That's a nice idea, but unfortunately it didn't help with this specific issue. (It's probably still a good idea for other reasons)

I replaced `recol` in that snippet with

```
cmd on-redraw :recol
```

and it has the same behavior WRT the `listen:` error message.

I also realized I should change the snippet to stop using `tput cols`, but that's a completely separate thing. (Readers: look up the current example in the docs for the best version)

**Reply by ilyagr** [reply]
I tried replacing `${{` with `%{{`. That resulted in the `listen:` message appearing in the `lf` status line instead of `stdout`. Perhaps that's what happened for you, but you have some other message output to `lf`'s status line later and replacing the error message?

**Reply by DusanLesan** [reply]
I have this working for me:
```
cmd on-redraw &{{
	if [ $lf_width -lt 90 ]; then
		lf -remote "send $id set preview false"
		lf -remote "send $id set ratios 1"
		$lf_cleaner
	elif [ $lf_width -lt 110 ]; then
		lf -remote "send $id set ratios 1:1"
		lf -remote "send $id set preview true"
	else
		lf -remote "send $id set preview true"
		lf -remote "send $id set ratios 1:2:3"
	fi
}}
```
But I have tried other prefixes as well and I did not experience issues. I have even tried your `recol` and it looked ok apart that I had issues with `$` prefix and built in commands

**Reply by joelim-work** [reply]
Hi @ilyagr @DusanLesan, I did some investigation and actually I can sort of reproduce this kind of issue if the server wasn't already started beforehand, using a more minimal config:

```sh
# the following sends a command to enable drawbox upon startup
# it fails if the server wasn't started beforehand already

# using echomsg also works too, if using a log file

$lf -remote "send $id set drawbox"
```

---

So this is caused by a startup race condition involving the following steps:

1. `lf` starts the server if it is not already started: https://github.com/gokcehan/lf/blob/e9faf966a035352eee958f50937279a4eaa416b2/main.go#L335-L337
2. `lf` spawns a separate thread to read commands from the server. This involves connecting to the server, registering its ID number using the `conn` server command, and then reading commands which will then be forwarded to the main thread: https://github.com/gokcehan/lf/blob/e9faf966a035352eee958f50937279a4eaa416b2/client.go#L128-L138
3. `lf` then reads the config file and executes any commands in it: https://github.com/gokcehan/lf/blob/e9faf966a035352eee958f50937279a4eaa416b2/app.go#L261-L273

The problem is that this only works as intended if the server was already started beforehand. If not, then what actually happens is that in step 2, the client thread will fail to connect to the server **and then sleep before retrying**. Since the `conn` command is only sent when the client thread manages to connect to the server, it means that it's possible for a `lf -remote` command in the config file to be sent to the server before that point, resulting in the `no such client id is connected` error message.

I'm not sure what's the best way to solve this problem though. I thought about blocking the rest of the initialization until the `conn` command is sent (by moving it to the main thread) - this works, but then the delay would be noticeable to users. Another way might to create some kind of `on-init` hook command, which is triggered internally when the `conn` command has been sent.

**Reply by joelim-work** [reply]
OK so I wrote a very quick implementation for this `on-init` hook command approach:

<details>
<summary>Click to see diff</summary>

```diff
diff --git a/client.go b/client.go
index 52360b4..aa2b825 100644
--- a/client.go
+++ b/client.go
@@ -125,7 +125,7 @@ func readExpr() <-chan expr {
 	ch := make(chan expr)
 
 	go func() {
-		duration := 1 * time.Second
+		duration := 1 * time.Millisecond
 
 		c, err := net.Dial(gSocketProt, gSocketPath)
 		for err != nil {
@@ -138,6 +138,7 @@ func readExpr() <-chan expr {
 		fmt.Fprintf(c, "conn %d\n", gClientID)
 
 		ch <- &callExpr{"sync", nil, 1}
+		ch <- &callExpr{"on-init", nil, 1}
 
 		s := bufio.NewScanner(c)
 		for s.Scan() {
diff --git a/eval.go b/eval.go
index df8b272..8e33582 100644
--- a/eval.go
+++ b/eval.go
@@ -594,6 +594,12 @@ func onFocusLost(app *app) {
 	}
 }
 
+func onInit(app *app) {
+	if cmd, ok := gOpts.cmds["on-init"]; ok {
+		cmd.eval(app, nil)
+	}
+}
+
 func onRedraw(app *app) {
 	if cmd, ok := gOpts.cmds["on-redraw"]; ok {
 		cmd.eval(app, nil)
@@ -1793,6 +1799,8 @@ func (e *callExpr) eval(app *app, args []string) {
 		onFocusGained(app)
 	case "on-focus-lost":
 		onFocusLost(app)
+	case "on-init":
+		onInit(app)
 	case "cmd-insert":
 		if len(e.args) == 0 {
 			return
```

</details>

And now the following config should work, even if the server isn't started beforehand:

```sh
set autoquit true

cmd on-init &{{
    lf -remote "send $id set drawbox"
}}
```

It's still a bit hacky though.

**Reply by DusanLesan** [reply]
I managed to reproduce the issue with a minimal setup. I also managed to reproduce it with `recol` now (thinking back, I did not check if I had the server left running). Even my `on-redraw` does not work as expected (it did not do anything, but I thought it did because my default is the same as the app default). Sorry for claiming otherwise.

I wouldn't mind having a hook to run after initialization is done. However, that might not be good enough for all cases. 
What if a user maps a command with remote actions and runs it as soon as they get into the UI? Would some kind of queue solve that, and is it safe to queue all commands? Is it possible for the server to know about clients being connected at the moment?

**Reply by joelim-work** [reply]
So I searched through the past issues and apparently this has been reported before in #495.

The main problem is that until the client successfully connects to the server and sends a `conn <id>` message to register itself, any kind of `lf -remote` command is not going to work. Regarding your idea of queuing commands until the client is ready, I don't think this is a proper fix - the user would still have to wait one second (or however long the registration takes) for anything involving remote commands to actually take effect. Another option is to have `lf -remote` fail if the client ID lookup fails, and retry based on the exit code, but it has the same issue of forcing the user to wait until the client has registered itself.

I still think that the main issue is the fact that the client takes too long to register itself, if it has to start up the server too. The only suggestions I have (none of them are particularly great) are:

- Decrease the sleep interval (currently one second). I don't know why the current value was chosen, but decreasing it will mean potentially more spam in the log file as the client fails to connect to the server. That might not be such a big problem however.
- Start the server beforehand (i.e. as part of your system startup). But then this suggests that the server should never quit and that `autoquit` shouldn't be enabled by default.
- Look at other alternatives for scripting dynamic configuration (e.g. using Lua), which don't require a server. But this is such a large change that it would probably be better to start from scratch (similar to Neovim), and then there is probably no point doing that either because you might as well use `yazi` instead.

**Reply by DusanLesan** [reply]
I did think queue is not a good solution, but I was not able to think of anything else. I am willing to run `lf` with reduced interval to see if I ever bump into any issues. As for logs about failed connection, can they be ignored first n times?

**Reply by joelim-work** [reply]
I did a very quick test, changing the initial retry sleep to 100 milliseconds (as opposed to 1 millisecond in the diff I posted earlier) seems to be sufficiently quick enough for *user-mapped commands* which call `lf -remote`, without spamming the log too much. I don't think a user press any keys that quickly immediately after opening `lf`.

But an `on-init` hook would still be required for anything that needs to be run automatically on startup.

**Reply by DusanLesan** [reply]
Do you want me to try it as well? I can keep ruining it for a while to see if there are any issues in regular use. Do you want to keep increasing duration exponentially?

**Reply by joelim-work** [reply]
Well if you take a look at the code, the initial sleep duration is one second when retrying, and then is doubled each time afterwards. But feel free to experiment with the initial duration and the `on-init` hook, and if it works I might consider submitting it as an actual PR.

But even then I don't know if I will end up changing the default for `autoquit` at this rate. So far there have only been two votes which is a very small sample size.

**Reply by DusanLesan** [reply]
@joelim-work I've been using [your implementation](https://github.com/gokcehan/lf/discussions/1767#discussioncomment-9943000) of `on-init` for a while now, and I have became dependent on it. Can we have that/something similar included in the `lf` codebase? 
My use case is to source additional config files after launch:
```bash
cmd on-init &{{
	[ -n "$TMUX" ] && lf -remote "send $id source $XDG_CONFIG_HOME/lf/tmux_conf"
}}
```

**Reply by joelim-work** [reply]
No worries, I have submitted this as a PR: #1838

So I have looked at the poll again, apparently it is now 6-0 in favor of enabling `autoquit`, I think I will go ahead with the changing the default after all, if there are no further objections.

**Reply by joelim-work** [reply]
@DusanLesan I have now merged both #1838 and #1839, could you please quickly verify that the new changes work for you?

**Reply by DusanLesan** [reply]
@joelim-work It looks good to me. Without the `autoquit` setting, `lf` quits by default, and the new hook works great. Thank you for making it

**Reply by ilyagr** [reply]
Sorry I lost track of this conversation. `on-init` sounds like a good solution, I will try it out soon, thank you for figuring it out!

**Reply by ilyagr** [reply]
https://github.com/gokcehan/lf/issues/1916 might end up having some related discussion.
